### DISCUSSION TITLE
Help: System.Relfection Exception. How do i resolve this issue? Can it be done with link.xml?

### DISCUSSION THREAD

**Discussion by ChimpSay** [applejag/Newtonsoft.Json-for-Unity#43] [Q&A] [answered]
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ExecutionEngineException: Attempting to call method 'Newtonsoft.Json.Utilities.CollectionWrapper`1[[System.UInt64, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]::.ctor
```
I have a class as such in a Unity Package, with an `.asmdef` named `Analytics.AppsFlyer.Runtime`
Is it possible to be fixed with a link.xml file or using AOTHelper is the best way?

```csharp
namespace Analytics.AppsFlyerPlatform
{
    public class CallFilter 
    {
        HashSet<ulong> set = new HashSet<ulong>();
    }
}
```
I tried many variation of `link.xml` file but none seems to work.
---
```xml
<linker>
	<assembly fullname="Analytics.AppsFlyer.Runtime" preserve="all" />
</linker>
```
---
```xml
<linker>
	<assembly fullname="Analytics.AppsFlyer.Runtime" preserve="all">
		<type fullname="Analytics.AppsFlyerPlatform">
			<type fullname="Analytics.AppsFlyerPlatform.CallFilter">
				<field name="set" />
				<field signature="System.Collections.Generic.HashSet`1&lt;System.UInt64&gt;" />
			</type>
		</type>
	</assembly>
</linker>
```

**Comment by ChimpSay** [comment]
By the way thanks for the awesome package, made it easy to use newtonsoft for unity 👍

**Comment by applejag** [comment, marked as answer]
Hi! Thanks for reporting, I'm eager to help :)

The exception is due to failing to deserialize your type, as you notice. However the issue does not lie in YOUR codebase, and that's why your link.xml attempts failed. To be specific, the errors says it fails to construct the following type from the `Newtonsoft.Json` assembly:

```cs
Newtonsoft.Json.Utilities.CollectionWrapper<ulong>
```

I strongly recommend solving this via AotHelper, however it can also be solved via link.xml. To solve via AotHelper you may add the following to a method inside a MonoBehaviour/ScriptableObject (any will do, not only one you're using):

```cs
AotHelper.EnsureList<ulong>();
```

You can find full specs of the AotHelper in the wiki page here: https://github.com/jilleJr/Newtonsoft.Json-for-Unity/wiki/Reference-Newtonsoft.Json.Utilities.AotHelper#aothelperensurelist

**Comment by ChimpSay** [comment]
Would you be so kind to show me how the link.xml would look like? I cant seem to figure that out.

**Comment by applejag** [comment]
Yes. However generics are not fully compatible with `link.xml`. The CollectionWrapper of type `ulong` is never explicitly referenced so telling the bytecode stripper to preserve it will do nothing for your use case. The AotHelper is not damaged by this at all as it forces the compilation of the type specified.

However for completeness sake, here's how to target the CollectionWrapper: (Note that you cannot target the specific generic type for preservation, but instead preserve ALL compiled generics for the CollectionWrapper)

```xml
<linker>
    <assembly fullname="Newtonsoft.Json">
        <type fullname="Newtonsoft.Json.Utilities.CollectionWrapper`1" preserve="all" />
    </assembly>
</linker>
```

I've written a small reference document for link.xml containing some of the common use cases. Have a look at the example in this page: https://github.com/jilleJr/Newtonsoft.Json-for-Unity/wiki/Reference-link.xml#the-method-element

**Comment by ChimpSay** [comment]
Ah gotcha. So i was preserving the wrong stuff. 
I think the easier solution would be just to use the AotHelper like you previously suggested.

Thanks once again 👍  
And i got it to work! 😄
